servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/TimeoutHttpRequesterFilter.java

```diff
@@ -100,4 +162,58 @@ public HttpExecutionStrategy influenceStrategy(final HttpExecutionStrategy strat
         // No influence since we do not block.
         return strategy;
     }
+
+    @FunctionalInterface
+    public interface TimeoutFromRequest extends Function<HttpRequestMetaData, Duration> {
+
+        /**
+         * Determine timeout duration, if present, from a request and/or apply default timeout durations.
+         *
+         * @param request the current request
+         * @return The timeout or null for no timeout
+         */
+        @Override
+        @Nullable
+        Duration apply(HttpRequestMetaData request);
+    }
+
+    /**
+     * Returns a function which converts a request to a timeout duration calculated based on the remaining time until
+     * the deadline in the context or, if absent, the provided default.
+     *
+     * <p>The timeout, if any, will be added to the context for additional client requests initiated within this context
+     *
+     * @param duration default timeout duration
+     * @return a function to produce a timeout based on the context deadline or specified default
+     */
+    public static TimeoutFromRequest useDefaultTimeout(@Nullable Duration duration) {
+        // the request is not considered
+        return (req) -> useContextDeadlineOrDefault(duration);
+    }
+
+    /**
+     * Returns timeout duration calculated based on the remaining time until the deadline in the context or, if absent,
+     * the provided default.
+     *
+     * <p>The timeout, if any, will be added to the context for additional requests initiated within this context
+     *
+     * @param defaultDuration default timeout duration or null for no timeout
+     * @return a timeout based on the context deadline or specified default (which may be null)
+     */
+    public static @Nullable Duration useContextDeadlineOrDefault(@Nullable Duration defaultDuration) {
+        Instant deadline = AsyncContext.get(HTTP_DEADLINE_KEY);
```

Comment:
I discovered that it didn't work when I disabled AsyncContext for debugging. I am not sure it needs to be documented because disabling AsyncContext isn't something that applications would do (I could be wrong in this assumption) and that when the ServiceTalks devs disable it they would understand that there might be behavioral differences.

Reply:
afaik AsyncContext [can be disabled by users too ](https://apple.github.io/servicetalk/servicetalk/SNAPSHOT/async-context.html#usage) if its not needed (ie. tracing), since its rather costly. Requiring it for timeouts (especially when this info isn't propagated to other requests 'yet') seems unreasonable. 

Me: _thinking out loud_
It looks to me that we need it only if we want to apply `remaining` duration for timeout on other actions upstream? Is there any other benefit? Can this functionality be user defined rather a filter default?